customer-service/src/main/kotlin/cz/kss/proj/customerservice/repository/CustomerRepositoryImpl.kt

```diff
@@ -0,0 +1,153 @@
+            null
+        } else {
+            Customer(
+                id = row["id"] as Long,
+                userName = row["user_name"] as String,
+                firstName = row["first_name"] as String,
+                lastName = row["last_name"] as String,
+                email = row["email"] as String,
+                additionalInformation = null,
+                address = null
+            )
+        }
+    }
+
+    override suspend fun delete(entity: Customer): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun save(entity: Customer): Customer {
+        return saveAdditionalInformation(saveAddress(this.saveCustomer(entity)))
+    }
+
+    suspend fun saveCustomer(entity: Customer): Customer {
+        return entity.id?.let { id ->
+            dbClient.sql(
+                """
+                        UPDATE customer SET user_name = :user_name,first_name = :first_name,last_name = :last_name, email = :email WHERE id = :id
+                    """.trimIndent()
+            )
+                .bind("user_name", entity.userName)
+                .bind("first_name", entity.firstName)
+                .bind("last_name", entity.lastName)
+                .bind("email", entity.email)
+                .fetch().first()
+                .mapNotNull { toEntity(it) }
+                .awaitSingle()
+        } ?: run {
+            dbClient.sql(
+                """
+                        INSERT INTO customer (user_name, first_name,last_name,email) VALUES (:user_name,:first_name,:last_name, :email)
+                    """.trimIndent()
+            )
+                .bind("user_name", entity.userName)
+                .bind("first_name", entity.firstName)
+                .bind("last_name", entity.lastName)
+                .bind("email", entity.email)
+                .filter { statement -> statement.returnGeneratedValues("id") }
+                .fetch().one()
+                .doOnNext { entity.id = it["id"] as Long }
+                .thenReturn(entity)
+                .awaitSingle()
+        } ?: throw IllegalStateException("Entity not saved")
+    }
+
+    suspend fun saveAddress(entity: Customer): Customer {
+        val t = addressRepository.save(entity.address!!)
+        entity.address = t
+        return entity
+    }
+
+    suspend fun saveAdditionalInformation(entity: Customer): Customer {
+        val t = additionalInformationRepository.save(entity.additionalInformation!!)
+        entity.additionalInformation = t
+        return entity
+    }
+
+    override suspend fun findById(id: Long): Customer? {
+        val traceIdContext = coroutineContext[TraceIdContext]
+        val correlationIdContext = coroutineContext[CorrelationIdContext]
+        logger.info("$traceIdContext $correlationIdContext find by id $id")
+        val currentMdc = MDC.getCopyOfContextMap()
+
+         return withContext(defaultCoroutineContext + MDCContext(currentMdc)) {
+            dbClient.sql(GET_CUSTOMER_BY_ID)
```

Comment:
This is the weakness point of MDC Context, I still do not why, but when you try to run dbClient without withContext context, it will be lost

Reply:
No clue right now. I'm not too happy about the necessity to populate the `currentMdc`. Let's review this later.